### PR TITLE
WEB-1140 manage user workspace membership services

### PR DESCRIPTION
**Pull request by cdriesler**
## Description & motivation

- We need repo/service functions for basic workspace role operations:
  - Create
  - Update
  - Remove
- We need to protect against the final admin of a workspace being removed
- We need to make sure relevant events are emitted by the service functions

## Changes:

- Adds repo functions and relevant integration tests
- Adds workspace role events and calls them in service functions

### CODE REVIEW

File: packages/server/modules/workspaces/repositories/workspaces.ts [hunk 1]
```diff
@@ -47,14 +48,76 @@ export const getWorkspaceRoleFactory =
...
+      .and.where('workspaceId', workspaceId)
+      .first()
+    const currentRole = await currentRoleQuery
+
+    if (!currentRole) {
+      return null
+    }
+
+    // Protect against removing last admin in workspace
+    const isUserLastAdmin = await isUserLastAdminFactory({ db })({
```
[the remark this answers is not in the document]
Comment (cdriesler) [reply packages/server/modules/workspaces/repositories/workspaces.ts:87]: I think it's also important to check, not just that the user is an admin, but that there are no other admins in the workspace. Is that maybe implied in another way from checking a given user's roles?

File: packages/server/modules/workspaces/repositories/workspaces.ts [hunk 2]
```diff
@@ -47,14 +48,76 @@ export const getWorkspaceRoleFactory =
...
+    return isUserAdmin && workspaceAdmins.length === 1
+  }
+
+export const deleteWorkspaceRoleFactory =
+  ({ db }: { db: Knex }): DeleteWorkspaceRole =>
+  async ({ userId, workspaceId }) => {
+    const workspacesAclTable = tables.workspacesAcl(db)
+
+    // Get current role
+    const currentRoleQuery = workspacesAclTable
```
[the remark this answers is not in the document]
Comment (cdriesler) [reply packages/server/modules/workspaces/repositories/workspaces.ts:76]: I think that's fair. Based on the roles listed I assumed a user would only ever have one role in a workspace.

Am I reading it right that you're suggesting dropping the `.where('role', 'workspace:admin')` bit and inspecting the list? That makes sense to me as well even if there's only ever one role. Is it strictly worse (performance wise?) to include this extra clause?

File: packages/server/modules/workspaces/services/workspaceRoleCreation.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    upsertWorkspaceRole,
+    emitWorkspaceEvent
+  }: {
+    upsertWorkspaceRole: UpsertWorkspaceRole
+    emitWorkspaceEvent: EmitWorkspaceEvent
+  }) =>
+  async ({ userId, workspaceId, role }: WorkspaceAcl): Promise<void> => {
+    await upsertWorkspaceRole({ userId, workspaceId, role })
+
+    // TODO: Should we return the final record from `upsert`, or `get`, instead of emitting args directly?
```
[the remark this answers is not in the document]
Comment (cdriesler) [reply packages/server/modules/workspaces/services/workspaceRoleCreation.ts:61]: This is something I plan to ask about next time we're all around. It looks like, in general, we _don't_ return the final result from upsert functions but this is a case where I'd find it useful.

File: packages/server/modules/workspaces/tests/integration/repositories.spec.ts [hunk 4]
```diff
@@ -8,11 +10,30 @@ import cryptoRandomString from 'crypto-random-string'
...
+const deleteWorkspaceRole = deleteWorkspaceRoleFactory({ db })
+const getWorkspaceRole = getWorkspaceRoleFactory({ db })
 const upsertWorkspaceRole = upsertWorkspaceRoleFactory({ db })
 
+const createAndStoreTestUser = async (): Promise<BasicTestUser> => {
+  const testId = cryptoRandomString({ length: 6 })
+
+  const userRecord: BasicTestUser = {
+    name: `test-user-${testId}`,
+    email: `test-user-${testId}@test.com`,
```
[the remark this answers is not in the document]
Comment (cdriesler) [reply packages/server/modules/workspaces/tests/integration/repositories.spec.ts:26]: roger, will do

**Comment by cdriesler** [comment]
@iainsproat took another swing, given your comments. I do like this much better lmk what you think

File: packages/server/modules/workspaces/repositories/workspaces.ts [hunk 5]
```diff
@@ -35,26 +38,92 @@ export const upsertWorkspaceFactory =
...
+  async ({ userId, workspaceId }) => {
+    // Protect against removing last admin in workspace
+    const workspaceRoles = await getWorkspaceRolesFactory({ db })({ workspaceId })
+
+    if (isUserLastWorkspaceAdmin(workspaceRoles, userId)) {
+      throw new Error('Cannot remove last admin in workspace.')
+    }
+
+    // Bail early if user has no role
+    const currentRole = workspaceRoles.find((role) => role.userId === userId)
```
Comment (gjedlicska) [inline packages/server/modules/workspaces/repositories/workspaces.ts:96]: instead of running a sql query against the db, inspecting the return value and based on a condition maybe run a second, you can issue a delete command with a returning '*' statement, to get back the deleted rows.

This way you are guaranteed to only run one sql statement.

https://knexjs.org/guide/query-builder.html#returning

File: packages/server/modules/workspaces/repositories/workspaces.ts [hunk 6]
```diff
@@ -35,26 +38,92 @@ export const upsertWorkspaceFactory =
...
   ({ db }: { db: Knex }): UpsertWorkspaceRole =>
   async ({ userId, workspaceId, role }) => {
+    // Verify requested role is valid workspace role
     const validRoles = Object.values(Roles.Workspace)
     if (!validRoles.includes(role)) {
       throw new Error(`Unexpected workspace role provided: ${role}`)
     }
 
+    // Protect against removing last admin in workspace
+    const workspaceRoles = await getWorkspaceRolesFactory({ db })({ workspaceId })
```
Comment (gjedlicska) [inline packages/server/modules/workspaces/repositories/workspaces.ts:118]: Now this the name of the function is lying to me. This func is more like `ensuringAnAdminRemainsAndUpsertsTheRoleFactory`

I think this is a lot cleaner as service level guarantee, its an application logic decision, that we do not allow workspaces without admin roles. The repo itself should not implement these guards. If we want that to be a strict dataschema guard, we should use database level guards instead..

File: packages/server/modules/workspaces/repositories/workspaces.ts [hunk 7]
```diff
@@ -35,26 +38,92 @@ export const upsertWorkspaceFactory =
...
+  return isUserAdmin && workspaceAdmins.length === 1
+}
+
+export const deleteWorkspaceRoleFactory =
+  ({ db }: { db: Knex }): DeleteWorkspaceRole =>
+  async ({ userId, workspaceId }) => {
+    // Protect against removing last admin in workspace
+    const workspaceRoles = await getWorkspaceRolesFactory({ db })({ workspaceId })
+
+    if (isUserLastWorkspaceAdmin(workspaceRoles, userId)) {
```
Comment (gjedlicska) [inline packages/server/modules/workspaces/repositories/workspaces.ts:91]: There could be cases, where we want to delete the role, even if the user is the last admin ie workspace deletion. ( If its not done via cascading deletes for some reason ). That way, this function becomes even more complex, cause based on an external logic, we might want to allow deletion after all. Its a lot better to have this check in the service implementation.

File: packages/server/modules/workspaces/services/workspaceRoleCreation.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (gjedlicska) [reply packages/server/modules/workspaces/services/workspaceRoleCreation.ts:61]: For me, an upsert is a complex conditional operation, cause if the value doesn't exist, you need to insert the full object. So allowing partial upsert needs a lot of guards to get right. I like full upserts more for that reason. If that is true, there must not be any differences between the args object and the stored value.

**Review by gjedlicska — CHANGES_REQUESTED** [review]

**Comment by cdriesler** [comment]
@gjedlicska understood and agreed on all counts. Repo functions are even less opinionated/think-y than I originally thought. Changes are in

File: packages/server/modules/workspaces/utils/isUserLastWorkspaceAdmin.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { WorkspaceAcl } from '@/modules/workspaces/domain/types'
+
+export const isUserLastWorkspaceAdmin = (
```
Comment (gjedlicska) [inline packages/server/modules/workspaces/utils/isUserLastWorkspaceAdmin.ts:3]: This is missing unit tests. I know the functionality is somewhat tested in the services, but in these cases i think it is required to fully test all the edge cases of this functionality.

File: packages/server/modules/workspaces/repositories/workspaces.ts [hunk 9]
```diff
@@ -35,21 +38,57 @@ export const upsertWorkspaceFactory =
...
-      .first()
+      .where({ workspaceId, userId })
+      .delete('*')
+
+    if (deletedRoles.length === 0) {
+      return null
+    }
 
-    return acl || null
+    return deletedRoles[0]
```
Comment (gjedlicska) [inline packages/server/modules/workspaces/repositories/workspaces.ts:85]: just make a note, that this is safe always, cause `workspaceId` and `userId` are composing a PK on the table, that ensures uniqueness. So there is always 0 or 1 entry returned here.

**Review by gjedlicska — CHANGES_REQUESTED** [review]

File: packages/server/modules/workspaces/services/workspaceRoleCreation.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    emitWorkspaceEvent: EmitWorkspaceEvent
+  }) =>
+  async ({
+    userId,
+    workspaceId
+  }: WorkspaceRoleDeleteArgs): Promise<WorkspaceAcl | null> => {
+    const workspaceRoles = await getWorkspaceRoles({ workspaceId })
+
+    if (isUserLastWorkspaceAdmin(workspaceRoles, userId)) {
+      throw new Error('Cannot remove last admin from a workspace.')
```
[the remark this answers is not in the document]
Comment (cdriesler) [reply packages/server/modules/workspaces/services/workspaceRoleCreation.ts]: done

**Review by gjedlicska — APPROVED** [review]
